src/partition.jl

```diff
@@ -5,6 +5,7 @@ mutable struct Partition
     cut_edges::Array{Int, 1}                    # of length(num_edges)
     dist_adj::SparseMatrixCSC{Int, Int}
     dist_nodes::Array{BitSet}
+    parent::Union{Partition, Nothing}           # optional parent partition
```

Comment:
Nifty! 

At first I was concerned about type instability (especially important because we pass Partition around literally everywhere...
but [this neat blogpost](https://julialang.org/blog/2018/08/union-splitting/)t put (and your @time s!) put that to rest.

Reply:
thanks for linking that blogpost!!! such an interesting read 🤔